pkg/proxy/iptables/proxier.go

```diff
@@ -552,6 +572,61 @@ func servicePortEndpointChainName(s proxy.ServicePortName, protocol string, endp
 	return utiliptables.Chain("KUBE-SEP-" + encoded[:16])
 }
 
+// getRemovedEndpointIPs returns the endpoint IPs that are missing in the new endpoints
+func getRemovedEndpoints(curEndpoints, newEndpoints []string) []string {
+	var result []string
+	epSet := sets.NewString(newEndpoints...)
+	for _, ep := range curEndpoints {
+		if epSet.Has(ep) {
+			result = append(result, ep)
+		}
+	}
+	return result
+}
+
+type endpointServicePair struct {
+	endpoint        string
+	servicePortName proxy.ServicePortName
+}
+
+// deleteEndpointConnection deletes all udp service connections to specified endpoint in the map to avoid stale endpoints black hole traffic
+func (proxier *Proxier) deleteEndpointConnection(connectionMap map[endpointServicePair]bool) {
+	for epSvcPair := range connectionMap {
+		if svcInfo, ok := proxier.serviceMap[epSvcPair.servicePortName]; ok && svcInfo.protocol == api.ProtocolUDP {
+			endpointIP := strings.Split(epSvcPair.endpoint, ":")[0]
+			glog.V(2).Infof("Attempt to delete connection to service %s with natted destination %s", svcInfo.clusterIP.String(), endpointIP)
+			err := proxier.execConntrackTool("-D", "--orig-dst", svcInfo.clusterIP.String(), "--dst-nat", endpointIP, "-p", "udp")
+			if err != nil {
+				glog.Errorf("conntrack tool return with error: %v", err)
+			}
+		}
+	}
+}
+
+// deleteServiceConnection use conntrack-tool to delete udp connection specified by service ip
+func (proxier *Proxier) deleteServiceConnection(svcIPs []string) {
+	for _, ip := range svcIPs {
+		glog.V(2).Infof("Attempt to delete udp connection to service %s", ip)
+		err := proxier.execConntrackTool("-D", "--orig-dst", ip, "-p", "udp")
+		if err != nil {
+			glog.Errorf("conntrack tool return with error: %v", err)
```

Comment:
Mark connections and delete them in a loop later sounds like a good idea. This also avoid noop conntrack deletes because there maybe no connections to the dead endpoint.